src/frequenz/sdk/timeseries/_formula_engine/_formula_engine_pool.py

```diff
@@ -0,0 +1,117 @@
+
+from frequenz.channels import Sender
+
+from ...actor import ChannelRegistry, ComponentMetricRequest
+from ...microgrid.component import ComponentMetricId
+from ._formula_generators._formula_generator import (
+    FormulaGenerator,
+    FormulaGeneratorConfig,
+)
+from ._resampled_formula_builder import ResampledFormulaBuilder
+
+if TYPE_CHECKING:
+    # Break circular import by enclosing these type hints in a `TYPE_CHECKING` block.
+    from .._formula_engine import (
+        FormulaReceiver,
+        FormulaReceiver3Phase,
+        _GenericEngine,
+        _GenericFormulaReceiver,
+    )
+
+
+class FormulaEnginePool:
+    """Creates and owns formula engines from string formulas, or formula generators.
+
+    If an engine already exist with a given name, it is reused instead.
+    """
+
+    def __init__(
+        self,
+        namespace: str,
+        channel_registry: ChannelRegistry,
+        resampler_subscription_sender: Sender[ComponentMetricRequest],
+    ) -> None:
+        """Create a new instance.
+
+        Args:
+            namespace: namespace to use with the data pipeline.
+            channel_registry: A channel registry instance shared with the resampling
+                actor.
+            resampler_subscription_sender: A sender for sending metric requests to the
+                resampling actor.
+        """
+        self._namespace = namespace
+        self._channel_registry = channel_registry
+        self._resampler_subscription_sender = resampler_subscription_sender
+        self._engines: dict[str, "FormulaReceiver|FormulaReceiver3Phase"] = {}
+
+    async def from_string(
+        self,
+        formula: str,
+        component_metric_id: ComponentMetricId,
+        nones_are_zeros: bool = False,
+    ) -> "FormulaReceiver":
+        """Get a receiver for a manual formula.
+
+        Args:
+            formula: formula to execute.
+            component_metric_id: The metric ID to use when fetching receivers from the
+                resampling actor.
+            nones_are_zeros: Whether to treat None values from the stream as 0s.  If
+                False, the returned value will be a None.
+
+        Returns:
+            A FormulaReceiver that streams values with the formulas applied.
+        """
+        channel_key = formula + component_metric_id.value
+        if channel_key in self._engines:
+            return self._engines[channel_key].new_receiver()
+
+        builder = ResampledFormulaBuilder(
+            self._namespace,
+            formula,
+            self._channel_registry,
+            self._resampler_subscription_sender,
+            component_metric_id,
+        )
+        formula_engine = await builder.from_string(formula, nones_are_zeros)
+        self._engines[channel_key] = formula_engine
+
+        return formula_engine.new_receiver()
+
+    async def from_generator(
+        self,
+        channel_key: str,
+        generator: "Type[FormulaGenerator[_GenericEngine]]",
```

Comment:
Is this type hint definition due to the circular import? I personally find it hacky but I guess there is not other workaround. Ideally we shouldn't have circular imports dependencies

Reply:
This is a Python problem,  it needs hacks to get around its circular imports problem.  And this hack is the recommended way to deal with it.

Quoting from: https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING

> The first type annotation must be enclosed in quotes, making it a “forward reference”, to hide the expensive_mod reference from the interpreter runtime. Type annotations for local variables are not evaluated, so the second annotation does not need to be enclosed in quotes.

Reply:
Oh I see, thanks for pointing that out